sound/soc/sof/ipc4-topology.c

```diff
@@ -1888,12 +1888,14 @@ static int sof_ipc4_prepare_gain_module(struct snd_sof_widget *swidget,
 	u32 out_ref_rate, out_ref_channels, out_ref_valid_bits;
 	int ret;
 
-	ret = sof_ipc4_init_input_audio_fmt(sdev, swidget, &gain->base_config,
-					    pipeline_params, available_fmt);
-	if (ret < 0)
-		return ret;
+	if (!swidget->prepared) {
```

Comment:
commit message

````
Add support for handling repeated ipc_prepare calls in the gain, src,
mixer and process type widgets. 
````

You should indicate why we need/want repeated ipc_prepare calls, it's far from obvious for someone like me who keeps forgetting what ipc_prepare is.

````
When the widget is prepared for first
time, the input audio format is initialized in the base_config. 
````
is this good or bad? And what would happen the next time this routine is called?

````
So
replace the use of the index returned from the call to
sof_ipc4_init_input_audio_fmt() to get the input audio to use the
base_config instead.
````

Whaaat?

````
 With repeated ipc_prepare calls, we only need to
update the pipeline params based on the chosen output format.
````
chosen when? the first time, last time?

````
This is only a preparatory step for now as a widget's ipc_prepare callback
be only invoked right now.
````

The grammar does not work for me, did you mean 'callback is only invoked once for now'?